File: modules/core/lib/rin/core/Disposable.cpp
#include "rin/core/Disposable.hpp"

void Disposable::OnDispose(bool manual)
{
}

void Disposable::Dispose(bool manual)
{
    if (_disposed) return;
    _disposed = true;
    OnDispose(manual);
}

bool Disposable::IsDisposed() const
{
    return _disposed;
}

Shared<Disposable> Disposable::GetShared()
{
    return this->shared_from_this();
}

Weak<Disposable> Disposable::GetWeak()
{
    return this->weak_from_this();
}

File: rin.Core/Disposable.cs
namespace rin.Core;

public abstract class Disposable : IAeroxDisposable
{
    public bool Disposed { get; set; }

    public string DisposeId { get; } = Guid.NewGuid().ToString();
    //public string DisposableId { get; private set; } = 

    public void Dispose()
    {
        Dispose(true);
        GC.SuppressFinalize(this);
    }

    protected abstract void OnDispose(bool isManual);

    protected virtual void Dispose(bool disposing)
    {
        if (Disposed) return;
        Disposed = true;
        OnDispose(disposing);
    }

    ~Disposable()
    {
        Dispose(false);
    }
    
    
}